### PR TITLE
Drug Categorization Code Added

### PR DESCRIPTION
**Pull request by Vedanth-Ramji**
The approach I used was to integrate the drug categorization functions in the normalizer itself. While adding the ARO columns, I also add 'OVERALL CATEGORY OF DRUG CLASS' and 'CONFERS RESISTANCE TO IMMEDIATE DRUG CLASS' columns and use my drug categorization functions to get the data to be displayed in those columns.

I have included the drug categorization in the 'run' function of the base normalizer class, as the drug categorization functions only need ARO numbers as parameters, hence after normalization, differences between database syntaxes wouldn't matter for categorization code.

### CODE REVIEW

File: argnorm/normalizers.py [hunk 1]
```diff
@@ -4,12 +4,20 @@
...
 MAPPING_TABLE_ARO_COL = 'ARO'
 TARGET_ARO_COL = 'ARO'
 
+# Column headings for drug categorization output
+IMMEDIATE_DRUG_CLASS_COL_HEADING = 'CONFERS RESISTANCE TO IMMEDIATE DRUG CLASS'
+DRUG_CLASS_CATEGORY_COL_HEADING = 'OVERALL CATEGORY OF DRUG CLASS'
+
 _ROOT = os.path.abspath(os.path.dirname(__file__))
 
-def get_data_path(path):
```
Comment (luispedro) [inline argnorm/normalizers.py:12]: I prefer the `get_data_path` name, tbh

File: argnorm/normalizers.py [hunk 2]
```diff
@@ -72,17 +98,17 @@ def get_aro_mapping_table(self):
         """
         Don't customize this unless you're using your own (not package built-in) reference data.
         """
-        df = pd.read_csv(get_data_path(f'{self.tool}_{self.database}_{self.mode}_ARO_mapping.tsv'), sep='\t', index_col=0)
+        df = pd.read_csv(get_data(f'{self.tool}_{self.database}_{self.mode}_ARO_mapping.tsv'), sep='\t', index_col=0)
         if self.tool != 'argsoap' or self.mode != 'orfs':
-            df[TARGET_ARO_COL] = df[TARGET_ARO_COL].map(lambda a: f'ARO:{int(a) if a == a else "nan"}') # a == a checks that a is not nan
```
Comment (luispedro) [inline argnorm/normalizers.py:77]: I think this is reverting some earlier refactorings

File: tests/test_arg_category.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+def test_get_immediate_drug_classes():
+    test_cases = ['ARO:3003938', 'ARO:3000014']
+
+    expected_output = [
+        [['ARO:3000008', 'penam'], ['ARO:0000032', 'cephalosporin'], ['ARO:0000044', 'cephamycin']],
+        [['ARO:3000008', 'penam'], ['ARO:0000004', 'monobactam'], ['ARO:3003706', 'penem'], ['ARO:0000032', 'cephalosporin'], ['ARO:3000637', 'ampicillin']]
+    ]
+
+    for i in range(len(test_cases)):
```
Comment (luispedro) [inline tests/test_arg_category.py:11]: Use `zip` instead of `range(len(...))`

File: tests/test_arg_category.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    test_cases = [
+        [['ARO:0000004', 'monobactam']],
+        [['ARO:3000157', 'rifamycin antibiotic']],
+        [['ARO:0000030', 'tigecycline']],
+        [['ARO:0000036', 'ciprofloxacin']]
+    ]
+
+    expected_output = [['beta-lactam antibiotic'], ['rifamycin antibiotic'], ['tetracycline antibiotic'], ['fluoroquinolone antibiotic']]
+
+    for i in range(len(test_cases)):
```
Comment (luispedro) [inline tests/test_arg_category.py:24]: same as above

File: tests/test_arg_category.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+def test_get_immediate_drug_classes():
+    test_cases = ['ARO:3003938', 'ARO:3000014']
+
+    expected_output = [
+        [['ARO:3000008', 'penam'], ['ARO:0000032', 'cephalosporin'], ['ARO:0000044', 'cephamycin']],
+        [['ARO:3000008', 'penam'], ['ARO:0000004', 'monobactam'], ['ARO:3003706', 'penem'], ['ARO:0000032', 'cephalosporin'], ['ARO:3000637', 'ampicillin']]
+    ]
+
+    for i in range(len(test_cases)):
+        assert sorted(drug_categorization.get_immediate_drug_classes(test_cases[i])) == sorted(expected_output[i]), drug_categorization.get_immediate_drug_classes(test_cases[i])
```
Comment (luispedro) [inline tests/test_arg_category.py:12]: Just write

```python
assert sorted(drug_categorization.get_immediate_drug_classes(test_cases[i])) == sorted(expected_output[i])
```

`pytest` is pretty good at generating informative output from this

File: argnorm/data/aro_ontology/aro.json
Comment (luispedro) [inline argnorm/data/aro_ontology/aro.json:1]: Are all these files necessary?

File: argnorm/drug_categorization.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    Parameters:
+        aro_num (str): ARO number. Needs to be in the form 'ARO:number'.
+
+    Returns:
+        drug_classes_list (list[list]): 
+            A two-dimensional list where each inner list represents a drug class.
+            Each inner list contains the ARO number and name of the drug class in that order. [ARO:number, name].
+    '''
+
+    GENE = ARO[aro_num]
```
Comment (luispedro) [inline argnorm/drug_categorization.py:23]: Avoid using uppercaps except for global constants

File: argnorm/drug_categorization.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    confers_resistance_to_drug_class = False
+    confers_resistance_to_antibiotic = False
+
+    if confers_resistance_to_drug_class == False or confers_resistance_to_antibiotic == False:
```
Comment (luispedro) [inline argnorm/drug_categorization.py:27]: This seems needlessly complicated.
```python
confers_resistance_to_drug_class = any(r.name == 'confers_resistance_to_drug_class' for r in GENE.relationships)
confers_resistance_to_antibiotic = any(r.name == 'confers_resistance_to_antibiotic' for r in GENE.relationships)
```

File: argnorm/drug_categorization.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    if confers_resistance_to_drug_class:
+        for drug_class in GENE.relationships[ARO.get_relationship('confers_resistance_to_drug_class')]:
+            drug_classes.append(drug_class)
+    
+    if confers_resistance_to_antibiotic:
+        for drug_class in GENE.relationships[ARO.get_relationship('confers_resistance_to_antibiotic')]:
+            drug_classes.append(drug_class)
+
+    drug_classes_list = []
+    for i in drug_classes:
```
Comment (luispedro) [inline argnorm/drug_categorization.py:46]: Nitpick, but `i` is normally for indices

File: argnorm/drug_categorization.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        drug_classes_list.append([i.id, i.name])
```
Comment (luispedro) [inline argnorm/drug_categorization.py:47]: This is unnecessary if you constructed the `drug_classes` list like this in the first place

File: argnorm/drug_categorization.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+def get_immediate_drug_classes(aro_num: str) -> list[list]:
+    '''
+    Description: Gets the drug classes to which a gene confers resistance to.
+    Only lists the drug class column in the CARD db.
+
+    Parameters:
+        aro_num (str): ARO number. Needs to be in the form 'ARO:number'.
+
+    Returns:
+        drug_classes_list (list[list]):
```
Comment (luispedro) [inline argnorm/drug_categorization.py:18]: A list of tuples seems slightly more idiomatic

**Review by luispedro — CHANGES_REQUESTED** [review]
Thanks @Vedanth-Ramji for this.

See all the single-line comments, but I also think that you committed several files that are not necessary. Also, I think this was not done from the `main` branch as it unwittingly reverts https://github.com/BigDataBiology/argNorm/commit/6baac446701862d4a7d9f70a1ea66b5b8d1e6a81 and https://github.com/BigDataBiology/argNorm/commit/19793d07b03cca854c3dca370c5a5dbed21b2740

File: argnorm/drug_categorization.py [hunk 11]
```diff
@@ -15,45 +15,34 @@ def get_immediate_drug_classes(aro_num: str) -> list[list]:
...
+    confers_resistance_to_drug_class = any(r.name == 'confers_resistance_to_drug_class' for r in gene.relationships)
+    confers_resistance_to_antibiotic = any(r.name == 'confers_resistance_to_antibiotic' for r in gene.relationships)
 
     drug_classes = []
 
     if confers_resistance_to_drug_class:
-        for drug_class in GENE.relationships[ARO.get_relationship('confers_resistance_to_drug_class')]:
-            drug_classes.append(drug_class)
+        for drug_class in gene.relationships[ARO.get_relationship('confers_resistance_to_drug_class')]:
+            drug_classes.append([drug_class.id, drug_class.name])
```
Comment (luispedro) [inline argnorm/drug_categorization.py:32]: This is not a `tuple`

Comment (Vedanth-Ramji) [reply argnorm/drug_categorization.py]: Yes, I missed it in the commit. I've added the tuples in this file with the changes to the smoke tests (i.e using the zip function) now.

File: tests/test_arg_category.py (hunk 3 above)
Comment (Vedanth-Ramji) [reply tests/test_arg_category.py]: I've used the zip function now as well

File: argnorm/normalizers.py (hunk 2 above)
Comment (Vedanth-Ramji) [reply argnorm/normalizers.py]: I've changed the code back to the way it is done in main now.

**Comment by luispedro** [comment]
The tests are now passing, but this commit has a lot of unnecessary material (remember that whatever you add in a commit will be in git's history forever). All the extra files should be removed before we can merge it

Also, the comment about `from_obo_library` is very relevant in this context: the `aro.obo` file is **not** used at all. I'd prefer to change that code to actually load from the copy of `aro.obo` that `argNorm` provides (although you need to take care of installing it and finding it in the right path!) but including a copy of `aro.obo` is a worst-of-both-worlds solution.

**Review by luispedro — APPROVED** [review]
Now LGTM

**Pull request merged**